src/main/java/com/openshift/internal/restclient/capability/server/Console.java

```diff
@@ -0,0 +1,186 @@
+ * Copyright (c) 2019 Red Hat, Inc. Distributed under license by Red Hat, Inc.
+ * All rights reserved. This program is made available under the terms of the
+ * Eclipse Public License v1.0 which accompanies this distribution, and is
+ * available at http://www.eclipse.org/legal/epl-v10.html
+ * 
+ * Contributors: Red Hat, Inc.
+ ******************************************************************************/
+
+package com.openshift.internal.restclient.capability.server;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import com.openshift.restclient.IClient;
+import com.openshift.restclient.ResourceKind;
+import com.openshift.restclient.capability.server.IConsole;
+import com.openshift.restclient.model.IConfigMap;
+import com.openshift.restclient.model.IProject;
+import com.openshift.restclient.model.IResource;
+
+public class Console implements IConsole {
+
+    public static final String CONFIGMAP_DATA_CONSOLE_URL = "consoleURL";
+    public static final String NAMESPACE_OPENSHIFT_CONFIG_MANAGED = "openshift-config-managed";
+    public static final String CONFIGMAP_CONSOLE_PUBLIC = "console-public";
+
+    private IClient client;
+
+    public Console(IClient client) {
+        this.client = client;
+    }
+
+    @Override
+    public boolean isSupported() {
+        return true;
+    }
+
+    @Override
+    public String getName() {
+        return this.getClass().getSimpleName();
+    }
+
+    @Override
+    public String getConsoleUrl() {
+        switch (client.getOpenShiftMajorVersion()) {
+        case 3:
+            return getOpenShift3ConsoleUrl();
+        case 4:
+            return getOpenShift4ConsoleUrl();
+        default:
+            return null;
+        }
+    }
+
+    @Override
+    public <R extends IResource> String getConsoleUrl(R resource) {
+        switch (client.getOpenShiftMajorVersion()) {
+        case 3:
+            return getOpenShift3ConsoleUrl(resource);
+        case 4:
+            return getOpenShift4ConsoleUrl(resource);
+        default:
+            return null;
+        }
+    }
+
+    private String getOpenShift3ConsoleUrl() {
+        return client.getBaseURL() + "/console";
+    }
+
+    private <R extends IResource> String getOpenShift3ConsoleUrl(R resource) {
+        StringBuilder builder = new StringBuilder(getOpenShift3ConsoleUrl());
+        String projectName = resource == null ? null : resource.getNamespaceName();
+        if (projectName != null) {
+            builder.append("/project/").append(projectName);
+        }
+        if (resource != null 
+                && !(resource instanceof IProject)) {
+            String consoleResourceUrl = getOpenShiftResourceURL(resource, 
+                    Arrays.stream(OpenShiftConsoleResourceUrls.values())
+                            .filter(resourceUrl -> resourceUrl.getVersion() ==  3));
+            if (consoleResourceUrl != null) {
```

Comment:
So if this is called for a resource type not supported by OpenShiftConsoleResourceUrls then null is returned but the builder has some kind of incomplete URL, is this on purpose ?

Reply:
if a resource type is not supported, then we should show the project overview. I added test (data points) that assert this.